network/framework/src/application/storage.rs

```diff
@@ -362,34 +459,82 @@ impl PeersAndMetadata {
     }
 }
 
-/// Returns the peer metadata for the given network
-fn get_peer_metadata_for_network<'a>(
-    peer_network_id: &'a PeerNetworkId,
-    peers_and_metadata: &'a mut RwLockWriteGuard<
-        HashMap<NetworkId, HashMap<AccountAddress, PeerMetadata>>,
-    >,
-) -> Result<&'a mut HashMap<AccountAddress, PeerMetadata>, Error> {
-    match peers_and_metadata.get_mut(&peer_network_id.network_id()) {
-        Some(peer_metadata_for_network) => Ok(peer_metadata_for_network),
-        None => Err(missing_network_metadata_error(
-            &peer_network_id.network_id(),
-        )),
+fn count_gauges<'a, T: Iterator<Item=(&'a NetworkId,&'a HashMap<AccountAddress,PeerMetadata>)>>(outer_iter: T) {
+    for (network_id, they) in outer_iter {
+        let mut inbound = 0;
+        let mut outbound = 0;
+        for (_peer_id, peer_metadata) in they.iter() {
+            match peer_metadata.connection_metadata.origin {
+                ConnectionOrigin::Inbound => {inbound += 1}
+                ConnectionOrigin::Outbound => {outbound += 1}
+            }
+        }
+        // info!("PeersAndMetadata count_gauges {} in={} out={}", network_id, inbound, outbound);
+        counters::connections(network_id, ConnectionOrigin::Inbound).set(inbound);
+        counters::connections(network_id, ConnectionOrigin::Outbound).set(outbound);
     }
 }
 
-/// A simple helper for returning a missing network metadata error
-fn missing_network_metadata_error(network_id: &NetworkId) -> Error {
-    Error::UnexpectedError(format!(
-        "No metadata was found for the given network: {:?}",
-        network_id
-    ))
-}
-
-/// A simple helper for returning a missing peer metadata error
+/// A simple helper for returning a missing metadata error
 /// for the specified peer.
-fn missing_peer_metadata_error(peer_network_id: &PeerNetworkId) -> Error {
+fn missing_metadata_error(peer_network_id: &PeerNetworkId) -> Error {
     Error::UnexpectedError(format!(
         "No metadata was found for the given peer: {:?}",
         peer_network_id
     ))
 }
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
+pub enum DisconnectReason {
```

Comment:
I'm still not entirely sure we need to keep this around. If we're adding generational support to the peers and metadata, surely it's cheap enough for applications to just keep polling the active peer set to be notified of changes?

Reply:
things _want_ events. but maybe I should look again because maybe _tests_ want events, and they can go fish if that's not what's right for the application code. But if things wind up polling and diffing to create events, we should just provide events. I think the event model is at least cleaner now that it's out of band with content messages and there's a separate Subscribe flow for connect/disconnect events.